### PR TITLE
Put dm.override_udev_sync_check back.

### PR DESCRIPTION
**Pull request by calavera**
- Warn when udev_sync is not enabled either because it's a statically linked binary or an old OS version.- 
- Update the link to the daemon storage options documentation.

This reverts commit 0a376291b2213699f986a7bca1cc8c4f4ed00f8d.

/cc @vbatts, @icecrime

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by icecrime** [comment]
LGTM (note for reviewers: it's just a revert, not the final thing for 1.7.1)

**Comment by icecrime** [comment]
For context: this is not only a revert, but also changes this [essential line](https://github.com/docker/docker/pull/14292/files#diff-4c195862cdadcaf026de87c0f6d7c313R1107) to make it a warning rather than an error.

This is something to avoid blocking users who were historically using devicemapper with a statically linked binary on Ubuntu, which we acknowledge is unreliable, but for which we don't have a solid, actionable alternative to provide right now.

The alternative will come in the form of dynamically linked binaries for all distros, which will take some time (but we're hoping to have them for 1.8.0). Once we do have those, we'll reconsider making devicemapper without udev sync an error, but right now it seems premature.

LGTM.

**Comment by jfrazelle** [comment]
SGTM

**Comment by crosbymichael** [comment]
LGTM

I would probably do things like this in two commits, one for the revert and one for the change so it's clean what was changed.

**Comment by calavera** [comment]
@crosbymichael divided in two commits

File: daemon/graphdriver/devmapper/deviceset.go [hunk 1]
```diff
@@ -1104,7 +1106,9 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 
 	// https://github.com/docker/docker/issues/4036
 	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
-		logrus.Warn("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see https://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option")
+		if !devices.overrideUdevSyncCheck {
+			logrus.Warn("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see https://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option")
```
Comment (vbatts) [inline daemon/graphdriver/devmapper/deviceset.go]: the override flag will be unused in this setup, but needed for the 1.8 release. They still need the warning regardless of the override check

File: daemon/graphdriver/driver.go [hunk 2]
```diff
@@ -170,31 +166,31 @@ func New(root string, options []string) (driver Driver, err error) {
...
-// it returns an error when there are several drivers scanned.
-func scanPriorDrivers(root string) (string, error) {
-	var priorDrivers []string
+// scanPriorDrivers returns an un-ordered scan of directories of prior storage drivers
+func scanPriorDrivers(root string) []string {
+	priorDrivers := []string{}
 	for driver := range drivers {
 		p := filepath.Join(root, driver)
-		if _, err := os.Stat(p); err == nil && driver != "vfs" {
+		if _, err := os.Stat(p); err == nil {
```
Comment (vbatts) [inline daemon/graphdriver/driver.go]: this vfs check needs to stay. that is not restoring prior behavior.

File: daemon/graphdriver/devmapper/deviceset.go (hunk 1 above)
Comment (calavera) [inline daemon/graphdriver/devmapper/deviceset.go]: what should we do with the check then?

File: daemon/graphdriver/driver.go (hunk 2 above)
Comment (calavera) [inline daemon/graphdriver/driver.go]: done

File: daemon/graphdriver/devmapper/deviceset.go (hunk 1 above)
Comment (icecrime) [inline daemon/graphdriver/devmapper/deviceset.go]: I think what @vbatts means is that the `if !devices.overrideUdevSyncCheck {` check here is not necessary. I don't think it makes a big difference and I believe we can keep it, but maybe I'm missing something.

Comment (vbatts) [inline daemon/graphdriver/devmapper/deviceset.go]: good question ...  `devicemapper.UdevSetSyncSupport(true)` is needed, but for 1.7.1 we could can skip/comment-out `devices.overrideUdevSyncCheck` test here

Comment (calavera) [inline daemon/graphdriver/devmapper/deviceset.go]: sounds good.

**Comment by icecrime** [comment]
LGTM

**Comment by vbatts** [comment]
LGTM

**Comment by cpuguy83** [comment]
I see you  are setting the override sync, but I don't see it being used by anything. Am I missing something?

**Comment by vbatts** [comment]
not missing anything. this is revert of behavior that 1.7.0 blocks for
many. It will be used in 1.8

**Comment by calavera** [comment]
Cherry picked into #14264

**Pull request merged**